src/psyclone/domain/lfric/metadata_to_arguments_rules.py

```diff
@@ -0,0 +1,614 @@
+        '''Adds an indirection dofmap for the provided function space and cma
+        operator when there is an apply cma kernel.
+
+        :param str function_space: the function space for this \
+            indirection dofmap.
+        :param cma_operator: the cma operator metadata associated with \
+            this indirection dofmap.
+        :type cma_operator: :py:class:`psyclone.domain.lfric.kernel.\
+            ColumnwiseOperatorArgMetadata`
+
+        '''
+
+    # pylint: disable=unidiomatic-typecheck
+    @classmethod
+    def _generate(cls, metadata):
+        '''Specifies which arguments appear in an argument list and their
+        ordering. Calls methods for each type of argument that can be
+        specialised by a child class for its particular need.
+
+        :param metadata: the LFRic kernel metadata.
+        :type metadata: \
+            py:class:`psyclone.domain.lfric.kernel.LFRicKernelMetadata`
+
+        :raises InternalError: if an unexpected mesh property is found.
+
+        '''
+        # All operator types require the cell index to be provided
+        if metadata.meta_args_get(
+                [OperatorArgMetadata, ColumnwiseOperatorArgMetadata]):
+            cls._cell_position()
+
+        # Pass the number of layers in the mesh unless this kernel is
+        # applying a CMA operator or doing a CMA matrix-matrix calculation
+        if metadata.kernel_type not in ["cma-apply", "cma-matrix-matrix"]:
+            cls._mesh_height()
+
+        # Pass the number of cells in the mesh if this kernel has a
+        # LMA operator argument.
+        # TODO issue #2074 this call should be used to replace the
+        # code that currently includes ncell3d for *every* operator it
+        # encounters (in _operator()).
+        # if metadata.meta_args_get(OperatorArgMetadata):
+        #     cls._mesh_ncell3d()
+
+        # Pass the number of columns in the mesh if this kernel operates on
+        # the 'domain' or has a CMA operator argument. For the former we
+        # exclude halo columns.
+        if metadata.operates_on == "domain":
+            cls._mesh_ncell2d_no_halos()
+        if metadata.meta_args_get(ColumnwiseOperatorArgMetadata):
+            cls._mesh_ncell2d()
+
+        if metadata.kernel_type == "inter-grid":
+            # Inter-grid kernels require special arguments.  The
+            # cell-map for the current column providing the mapping
+            # from the coarse to the fine mesh.
+            cls._cell_map()
+
+        # For each argument in the order they are specified in the
+        # kernel metadata, call particular methods depending on what
+        # type of argument we find (field, field vector, operator or
+        # scalar). If the argument is a field or field vector and also
+        # has a stencil access then also call appropriate stencil
+        # methods.
+        const = LFRicConstants()
+        for meta_arg in metadata.meta_args:
+
+            if type(meta_arg) in [
```

Comment:
I could use the class hierarchy here to do `isinstance(meta_arg, FieldArgMetadata)` for the first if test, then keep `type(meta_arg) in [FieldArgMetadata, InterGridArgMetadata]` for the second if test and then isinstance(meta_arg, (FieldVectorArgMetadata, InterGridVectorArgMetadata))` for the 3rd if, if you think that is more readable code?

Reply:
Again, it's whether having to disable a pylint warning is telling us that the code would be better in a different form. Perhaps, as you say, we're misusing inheritance between the metadata classes? Anyway, what you have is perfectly readable so we can leave it as is.

Reply:
Yes, I think @sergisiso agrees with you as well. When I explained the need to use `type()` rather than `isinstance()` in certain comparisons due to the metadata class hierarchy he wondered whether I'd used the right hierarchy in the first place. In my defence, it makes sense from a metadata hierarchy point of view. `FieldMetadata` is `ScalarMetadata` with an additional `function_space`. The problem is that it does not make sense from an actual `Field` and `Scalar` point of view, which is how we are making use of the metadata classes in practice when we want to pull out certain types. I should probably revisit this given that you both don't like it. I think the only solution would be to remove the inheritance as all datatypes are different. We could then replace type with isinstance. A quick google says you can use methods from other classes in Python 3 so this should be feasible without massive changes to the implementation.

Reply:
I think a new issue for this then please.